app/src/main/java/com/kelvin/bdaypost/birthday/add/AddBirthdayFragment.kt

```diff
@@ -0,0 +1,159 @@
+
+        // create VM
+        addBdayVM = activity?.run {
+            ViewModelProvider(this, AddBirthdayViewModel.AddBirthdayVMFactory())
+                .get(AddBirthdayViewModel::class.java)
+        }
+        // Add VM observers
+        addBdayVM?.run {
+            this.validContactForm.observe(
+                this@AddBirthdayFragment.viewLifecycleOwner,
+                Observer { observeContactState(it) }
+            )
+
+            this.addContactResult.observe(
+                this@AddBirthdayFragment.viewLifecycleOwner,
+                Observer { observeAddContact(it) }
+            )
+        }
+
+
+        // configure date picker for birthday spinner
+        context?.let { addDatePicker(it) }
+        // add btn listener for confirm saving birthday
+        addBdayBinding.btnAddBdayAddContact.setOnClickListener { saveBday() }
+
+        // debug address for easy testing
+        if (BuildConfig.DEBUG) {
+            addBdayBinding.inputAddBdayContactName.setText("Kelvin")
+            addBdayBinding.inputAddBdayContactAddr.setText("742 evergreen terrace")
+        }
+    }
+
+    /**
+     * Populates the month birthday picker with array data onViewCreated
+     * Adds listener to change the date to the appropriate range based on month selected
+     */
+    private fun addDatePicker(context: Context) {
+        addBdayBinding.spinAddBdayMonth.adapter =
+            ArrayAdapter(context, android.R.layout.simple_spinner_item, Month.monthList)
+
+        addBdayBinding.spinAddBdayMonth.onItemSelectedListener = object: AdapterView.OnItemSelectedListener {
+            override fun onNothingSelected(parent: AdapterView<*>?) {
+                Timber.d("Nothing selected")
+            }
+
+            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
+                changeBdayDateRange(Month.monthList[position])
+            }
+        }
+    }
+
+    /**
+     * Watches for add birthday saving results
+     */
+    private fun observeAddContact(contactAddedResult: AddBirthdayViewModel.ContactAddedResult?) {
+        contactAddedResult?.let {
+            it.success?.let { contact -> this.context?.showToast("${contact.contactInfo.name} saved!")}
+            it.error?.let { errRes -> this.context?.showToast(errRes) }
+        }
+    }
+
+    /**
+     * Watches for valid add contact fields prior to saving
+     */
+    private fun observeContactState(contactState: AddBirthdayViewModel.ContactFormState?) {
+        contactState?.let {
+            contactState.nameError?.let {
+                addBdayBinding.inputAddBdayContactName.error = this.getString(it) }
+            contactState.addrError?.let {
+                addBdayBinding.inputAddBdayContactAddr.error = this.getString(it) }
+        }
+    }
+
+    /**
+     * Validates the fields and then saves the contact
+     */
+    private fun saveBday() {
+        val name = addBdayBinding.inputAddBdayContactName.text.toString()
```

Comment:
I like how a lot of the business logic is in the view model class, I wonder if we can move the `if` block there too so that this function has even less business logic. Like 
```
val name = addBdayBinding.inputAddBdayContactName.text.toString()
val addr = addBdayBinding.inputAddBdayContactAddr.text.toString()
val bdayMonth = addBdayBinding.spinAddBdayMonth.selectedItem.toString()
val bdayDate = addBdayBinding.spinAddBdayDate.selectedItem.toString()
saveBdayIfValid(name, addr, bdayMonth, bdayDate)
```
and the `saveBdayIfValid()` can validate the contact form without exposing that to the fragment.